_posts/2020-05-04-strangler-pattern.md

```diff
@@ -0,0 +1,107 @@
+
+![Strangler Pattern](/img/strangler-pattern/strangler-pattern.png)
+
+**Stage 1**: Identify the system boundary you want to re-architect.
+
+**Stage 2**: Create a [facade](https://en.wikipedia.org/wiki/Facade_pattern).
+
+* The facade ensures clients of the system are unaffected by the re-architecture activity, and should therefore match an existing system API.
+* The API for the new service should be carefully designed to ensure backward compatability with the legacy API.  If you want to make breaking changes to the API for the new service, that should be performed once re-architecture is completed.
+
+**Stage 3**: Re-architect.
+
+* Implement your new service with a compatible.
+* Dark launch the new service: to support a controlled rollout, and tee traffic between the new and legacy implementations.
+* Return responses from the legacy implementation.  
+* Capture, or discard, responses from the new implementation.  This enables you to monitor the behaviour of the new implementation to ensure it is fit for purpose.
+* This also provides a mechanism to support continuous deployment: during development, deploy partial implementations of the new service, to gain confidence and mitigate risk without compromising system behaviour
+
+**Stage 4**: Cutover.
+* Once you are confident that the new implementation meets your needs, cutover production traffic so requests and responses are served from the new implementation.
+* Retain the legacy implementation to allow for easy rollback if required.
+
+**Stage 5**: Clean up the mess.
+* Once the new system is established delete the legacy code that has been replaced.
+
+**Stage 6**: Migrate clients to the new API and remove the strangler facade
+* This is likely to require breaking changes to clients, and should be managed accordingly.
+* To manage the transition, if you have a large number of clients, expose the new API whilst maintaining the strangler facade.  
+    * Monitor usage of the strangler facade and proactively engage with clients to help them migrate.  
+    * Once the strangler facade is no longer being used, it can be safely removed.
+
+# Considerations
+
+* Break up the system based on new service boundaries, or integration layers.
+* Removal of the strangler facade, and any unused legacy code is an important step to simplify the overall solution.
+* Adequate monitoring and observability is requried to validate the new service is fit for purpose.
+
+# Case Study: Infrastructure Modernisation
+
+The strangler pattern doesn't just apply to software architecture, but can be used as part of your cloud migration journey.
+
+We helped one client replatform from an on-premise application servers to containers running on AWS Fargate.  The application consisted of a decoupled, event driven, architecture.
```

Comment:
Need more details of their architecture. 
Explain the horrible complexity. 
Explain some of the problems, some of the reasons to migrate. 
What was the point?

Reply:
You mention "each sservice"  in the next para but don't mention the original app had 4 chunky services. And the diagram is a bit unclear.